File: labs/lab1.3/csharp/MyString.cs
// Рядки: значення рядка (символи), конструктор
// за умовчанням, конструктор з параметрами,
// конструктор копіювання, метод обчислення
// довжини рядка, метод отримання значення
// рядка, перевантаження операторів
// додавання, вирахування. Метод вирахування
// повинен передбачати можливість роботи з
// кількома рядками

namespace Program;

class MyString {
    public string Value {get;set;}

    public MyString() => this.Value = "";
    public MyString(string text) => this.Value = text;
    public MyString(char[] text) => this.Value = new String(text);
    public MyString(MyString other) => this.Value = other.Value;

    public int Length() => this.Value.Length;
    public override string ToString() => this.Value;
    public override bool Equals(object obj) {
        if(ReferenceEquals(this,obj)) return true;
        if(obj is MyString other) return this == other;
        return false;
    }
    public override int GetHashCode() => this.Value.GetHashCode();

    public static MyString operator +(MyString a, MyString b) => new MyString(a.Value + b.Value);
    public static MyString operator -(MyString a, MyString b) => new MyString(a.Value.Replace(b.Value,""));
    public static MyString operator -(MyString a, string b) => new MyString(a.Value.Replace(b,""));
    public static MyString operator -(MyString a, char b) => new MyString(a.Value.Replace(b,'\0'));
    public static bool operator ==(MyString a, MyString b) => a.Value == b.Value;
    public static bool operator !=(MyString a, MyString b) => a.Value != b.Value;
}

File: labs/lab1.3/cpp/myString.cpp
#include "myString.h"

myString::myString() {this->Value = "";}
myString::myString(const string &text) {this->Value = text;}
myString::myString(char text[]) {this->Value = string(text);}
myString::myString(const myString &other) {this->Value = other.Value;}

int myString::Length() {return this->Value.length();}

myString myString::operator+(const string &other) {return myString(this->Value + other);}
myString myString::operator+(const myString &other) {return myString(this->Value + other.Value);}
myString myString::operator+(const char *other) {return myString(this->Value + string(other));}
myString myString::operator+(char other) {return myString(this->Value + string(1,other));}

myString myString::operator-(const string &other) {
    string result = this->Value;
    size_t pos = result.find(other);
    while(pos != string::npos) {
        result.erase(pos, other.length());
        pos = result.find(other);
    }
    return myString(result);
}
myString myString::operator-(const myString &other) {
    return (*this) - other.Value;
}
myString myString::operator-(const char *other) {
    return (*this) - string(other);
}
myString myString::operator-(char other) {
    return (*this) - string(1,other);
}

bool myString::operator==(const myString &other) {return this->Value == other.Value;}
bool myString::operator!=(const myString &other) {return this->Value != other.Value;}